### PR TITLE
Windows: First part of CI tests (docker run)

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard jhoward@microsoft.com

@jfrazelle @cpuguy83.  This is the first of several updates to integration-cli to start enabling test-integration-cli to run against the Windows daemon. It addresses many of the tests in docker_cli_run_test.go (ie `TestRun*`). These can't yet be run under CI (still more infrastructure required), but can locally. I use an interim Windows `busybox` image which I build from http://github.com/jhowardmsft/busybox. This image will change though, as it's 32-bit and hence can't run on nano server.

### CODE REVIEW

**Comment by calavera** [comment]
I'm very excited about this!

File: integration-cli/docker_test_vars.go [hunk 1]
```diff
@@ -34,8 +34,19 @@ var (
...
 	// a version call to the daemon and examining the response header.
 	daemonPlatform string
+
+	// daemonDefaultImage is the name of the default image to use when running
+	// tests. This is platform dependent.
+	daemonDefaultImage string
 )
 
+// WindowsBaseImage is the name of the base image for Windows testing
+const WindowsBaseImage = "windowsservercore"
```
Comment (jfrazelle) [inline integration-cli/docker_test_vars.go]: can this be in a pretty `()` along with the other

Comment (jfrazelle) [inline integration-cli/docker_test_vars.go]: like 
```
const (
)
```

**Comment by jhowardmsft** [comment]
@calavera Me too! Stopping Windows regressions is a major step forward :)

File: integration-cli/docker_test_vars.go (hunk 1 above)
Comment (jhowardmsft) [inline integration-cli/docker_test_vars.go]: @jfrazelle Of course. Updated :)

**Comment by jhowardmsft** [comment]
guys - any chance of getting this merged? I want to build on it for verifying the forthcoming volume support on Windows :)

**Comment by jhowardmsft** [comment]
And rebased

File: integration-cli/docker_cli_run_test.go [hunk 2]
```diff
@@ -404,8 +446,15 @@ func (s *DockerSuite) TestRunVolumesFromSymlinkPath(c *check.C) {
...
 
 func (s *DockerSuite) TestRunExitCode(c *check.C) {
-	testRequires(c, DaemonIsLinux)
-	_, exit, err := dockerCmdWithError("run", "busybox", "/bin/sh", "-c", "exit 72")
+	var (
+		exit int
+		err  error
+	)
+	if daemonPlatform == "windows" {
+		_, exit, err = dockerCmdWithError("run", WindowsBaseImage, "cmd", "/s", "/c", "exit 72")
```
Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: busybox doesn't work for this test?

File: integration-cli/docker_cli_run_test.go [hunk 3]
```diff
@@ -521,6 +591,7 @@ func (s *DockerSuite) TestRunEnvironment(c *check.C) {
 		}
 	}
 	sort.Strings(actualEnv)
+	fmt.Println(actualEnv)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: Remove

File: integration-cli/docker_cli_run_test.go [hunk 4]
```diff
@@ -1388,7 +1507,7 @@ func (s *DockerSuite) TestRunCopyVolumeContent(c *check.C) {
 }
 
 func (s *DockerSuite) TestRunCleanupCmdOnEntrypoint(c *check.C) {
-	testRequires(c, DaemonIsLinux)
+	//testRequires(c, DaemonIsLinux)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: Remove

File: integration-cli/docker_cli_run_test.go [hunk 5]
```diff
@@ -1606,7 +1752,7 @@ func (s *DockerSuite) TestRunBindMounts(c *check.C) {
 // Ensure that CIDFile gets deleted if it's empty
 // Perform this test by making `docker run` fail
 func (s *DockerSuite) TestRunCidFileCleanupIfEmpty(c *check.C) {
-	testRequires(c, DaemonIsLinux)
+	//testRequires(c, DaemonIsLinux)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: Remove

File: integration-cli/docker_cli_run_test.go [hunk 6]
```diff
@@ -1958,12 +2130,25 @@ func (s *DockerSuite) TestRunExposePort(c *check.C) {
...
-	c.Assert(err, check.NotNil)
+
+	// Windows and Linux are different here by architectural design. Linux will
+	// fail to start the container, so an error is expected. Windows will
+	// successfully start the container, and once started attempt to execute
+	// the command which will fail.
+	if daemonPlatform == "windows" {
+		// Wait for it to exit.
+		waitExited(cID, 30)
+		c.Assert(err, check.IsNil)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: But shouldn't the client expect an error here? If not, this is broken.

File: integration-cli/docker_cli_run_test.go [hunk 7]
```diff
@@ -3092,11 +3342,16 @@ func (s *DockerSuite) TestRunLoopbackOnlyExistsWhenNetworkingDisabled(c *check.C
 
 // Issue #4681
 func (s *DockerSuite) TestRunLoopbackWhenNetworkDisabled(c *check.C) {
-	testRequires(c, DaemonIsLinux)
-	dockerCmd(c, "run", "--net=none", "busybox", "ping", "-c", "1", "127.0.0.1")
+	//testRequires(c, DaemonIsLinux)
```
Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: Remove

Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: This test is funny...
There is another test `TestRunContainerNetwork` which tests the exact same thing, but without `--net=none`... I think the other test is probably not valid... just taking a note of this.
ping @mavenugo.

**Comment by cpuguy83** [comment]
A couple of minor comments.

I do have some concerns about the test complexity with checking for the daemon type, but I understand it is needed at this point and getting tests running on Windows is paramount.

File: integration-cli/docker_cli_run_test.go (hunk 2 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: No - it appears to be a bug in the busybox image I found. Our new hire is looking at other bugs in it and making sure there's a 64-bit image available for nanoserver testing. This is a workaround though.

File: integration-cli/docker_cli_run_test.go (hunk 4 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: Oops, removed.

File: integration-cli/docker_cli_run_test.go (hunk 5 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: Ditt

File: integration-cli/docker_cli_run_test.go (hunk 6 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: See my block comment above :) Windows will successfully start the container and fail when the program being launched in the container cannot be found. This is different to Linux. There's nothing we can do here from a docker perspective - it's inherent by design.

File: integration-cli/docker_cli_run_test.go (hunk 7 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: Removed.

File: integration-cli/docker_cli_run_test.go (hunk 6 above)
Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: Ah, I guess so since it is just a start w/o an attach.

**Comment by jhowardmsft** [comment]
Yeah, it's unfortunate, but I'm not sure of an easier way at this time. Comments addressed. Forced push imminent.

File: integration-cli/docker_cli_run_test.go (hunk 3 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: Removed.

**Comment by calavera** [comment]
LGTM

File: integration-cli/docker_cli_run_test.go [hunk 8]
```diff
@@ -620,11 +696,16 @@ func (s *DockerSuite) TestRunEnvironmentOverride(c *check.C) {
...
+	if daemonPlatform == "windows" {
+		// Windows busybox does not have ping. Use built in ping instead.
+		dockerCmd(c, "run", WindowsBaseImage, "ping", "-n", "1", "127.0.0.1")
+	} else {
+		dockerCmd(c, "run", "busybox", "ping", "-c", "1", "127.0.0.1")
+	}
 }
 
 func (s *DockerSuite) TestRunNetHostNotAllowedWithLinks(c *check.C) {
+	// TODO Windows: --link is not yet functional.
```
Comment (tiborvass) [inline integration-cli/docker_cli_run_test.go]: @jhowardmsft fyi, not 100% sure if you even need to get `--link` working on windows since we're moving towards `docker network` and `docker service`. /cc @mavenugo

File: integration-cli/docker_cli_run_test.go [hunk 9]
```diff
@@ -1555,19 +1692,26 @@ func (s *DockerSuite) TestRunWithBadDevice(c *check.C) {
...
-	testRequires(c, DaemonIsLinux)
 	name := "entrypoint"
-	out, _ := dockerCmd(c, "run", "--name", name, "--entrypoint", "/bin/echo", "busybox", "-n", "foobar")
 
-	expected := "foobar"
+	// Note Windows does not have an echo.exe built in.
+	var out, expected string
+	if daemonPlatform == "windows" {
+		out, _ = dockerCmd(c, "run", "--name", name, "--entrypoint", "cmd /s /c echo", "busybox", "foobar")
+		expected = "foobar\r\n"
```
Comment (tiborvass) [inline integration-cli/docker_cli_run_test.go]: @jhowardmsft really nbd, it's just out of curiosity, isn't there an `echo -n` equivalent in windows so that expected can be `foobar` for both?

File: integration-cli/docker_utils.go [hunk 10]
```diff
@@ -1405,6 +1405,12 @@ func waitRun(contID string) error {
 	return waitInspect(contID, "{{.State.Running}}", "true", 5)
 }
 
+// waitExited will wait for the specified container to state exit, subject
+// to a maximum time limit in seconds supplied by the caller
+func waitExited(contID string, seconds int) error {
```
Comment (tiborvass) [inline integration-cli/docker_utils.go]: should take a `time.Duration` instead of an `int`

File: integration-cli/docker_cli_run_test.go [hunk 11]
```diff
@@ -2165,9 +2354,11 @@ func (s *DockerSuite) TestRunTtyWithPipe(c *check.C) {
 
 		expected := "cannot enable tty mode"
 		if out, _, err := runCommandWithOutput(cmd); err == nil {
+			fmt.Println("out1=", out)
```
Comment (tiborvass) [inline integration-cli/docker_cli_run_test.go]: left over debug?

File: integration-cli/docker_cli_run_test.go [hunk 12]
```diff
@@ -1958,12 +2127,25 @@ func (s *DockerSuite) TestRunExposePort(c *check.C) {
...
 	_, _, err := dockerCmdWithError("start", cID)
-	c.Assert(err, check.NotNil)
+
+	// Windows and Linux are different here by architectural design. Linux will
+	// fail to start the container, so an error is expected. Windows will
+	// successfully start the container, and once started attempt to execute
+	// the command which will fail.
+	if daemonPlatform == "windows" {
+		// Wait for it to exit.
+		waitExited(cID, 30)
```
Comment (tiborvass) [inline integration-cli/docker_cli_run_test.go]: 30 seconds seems like a long time

File: integration-cli/docker_cli_run_test.go (hunk 2 above)
Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: Can we get a todo to make this work with busybox?

File: integration-cli/docker_cli_run_test.go (hunk 6 above)
Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: I'm not sure where the design discussions happened around this or if it is a limitation in Windows, but this really seems like a problem and a huge inconsistency in the API.

File: integration-cli/docker_cli_run_test.go (hunk 8 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: Do you want me to change the comment?

File: integration-cli/docker_cli_run_test.go (hunk 9 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: Not from the inbox echo unfortunately

File: integration-cli/docker_utils.go (hunk 10 above)
Comment (jhowardmsft) [inline integration-cli/docker_utils.go]: I used the same signature as the existing waitInspect.... Do you want both changed and the callers?

File: integration-cli/docker_cli_run_test.go (hunk 11 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: Well spotted. Yes, removing this and the one a few lines lower down.

File: integration-cli/docker_cli_run_test.go (hunk 12 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: On SSD, server core can take up to 6 seconds. On spinning, that will be more (circa 10 I suspect, but I only have SSDs ;) ). I was trying to ensure the test was reliable. If it takes 30 seconds, that should be indicative of a failure - I wanted to set something high enough that it -should- never reach.

File: integration-cli/docker_cli_run_test.go (hunk 8 above)
Comment (tiborvass) [inline integration-cli/docker_cli_run_test.go]: @jhowardmsft yes, also feel free to contact @mavenugo (madhu on IRC) for the networking and service discovery work.

Comment (mavenugo) [inline integration-cli/docker_cli_run_test.go]: @tiborvass +1. link will be deprecated in favor of CNM. `service discovery` needs more work and that is exactly where we should put our efforts on.

File: integration-cli/docker_utils.go (hunk 10 above)
Comment (tiborvass) [inline integration-cli/docker_utils.go]: sure it'd be easier to read

File: integration-cli/docker_cli_run_test.go (hunk 8 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: So what do you want me to change the comment to? Bearing in mind this is an existing Linux test regardless of futures...

Comment (tiborvass) [inline integration-cli/docker_cli_run_test.go]: Just remove the comment and let's keep it linux specific.

File: integration-cli/docker_cli_run_test.go (hunk 2 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: Why? (Sorry, not sure really that's relevant as most people won't in reality be testing busybox on Windows, but *will* be using cmd and more specifically running containers checking exit codes from cmd. IMO this is a more realistic test on Windows).

Comment (tiborvass) [inline integration-cli/docker_cli_run_test.go]: @jhowardmsft i think he was just referring to your "workaround" comment: https://github.com/docker/docker/pull/16312/files#r39983870

From that, it seems like once the new hire debugged the issue, we could update this piece of code, hence the TODO.

File: integration-cli/docker_cli_run_test.go (hunk 8 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: As I'm going through all the CI files, I'm trying to explicitly comment each test as to why it's not relevant on Windows to make it easier for future maintenance. I'll update the comment to say it's Linux specific due to --link not supported on Windows and this will be deprecated in favour of CNM. OK?

File: integration-cli/docker_cli_run_test.go (hunk 2 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: I think you're missing my point :) How many people on Windows are actually going to be using the busybox image in the real world. Yet how many are going to be using cmd? I have no concern filing an internal bug against Darren to look at why it fails though. Make sense?

Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: Indeed.
Aren't we really testing here that the daemon sets the expected exit code and the client exits with said exit code rather than a specific feature of cmd or sh?
Basically the reason is to keep the test as simple as possible, if we can make it work with the same command on both platforms that's ideal.

File: integration-cli/docker_cli_run_test.go (hunk 8 above)
Comment (tiborvass) [inline integration-cli/docker_cli_run_test.go]: @jhowardmsft Sure that works

File: integration-cli/docker_cli_run_test.go (hunk 2 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: I give in :) Sure, I'll add the comment, but I honestly don't really agree.

**Comment by jhowardmsft** [comment]
All comments (to date, lol....) addressed.

**Comment by tiborvass** [comment]
LGTM

File: integration-cli/docker_cli_run_test.go (hunk 6 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: I kind of go back to the reality statement that Windows does not behave the same as Linux and is not Linux with a different API set. They are completely different platforms. We cannot tell up front whether an application will actually start in a container without starting the container first.

Comment (tiborvass) [inline integration-cli/docker_cli_run_test.go]: I personally think it's fine, and I don't see what else we can do about it. On windows `start` should error out if docker encountered a problem (like inexistant container name, or something went wrong in the internals prior to starting the container). The container's exit code itself can be retrieved with `wait` and that's consistent everywhere.

**Comment by cpuguy83** [comment]
```
FAIL: docker_cli_run_test.go:2475: DockerSuite.TestRunRestartMaxRetries

docker_cli_run_test.go:2484:
    c.Fatal(err)
... Error: condition ""false true" == "false false"" not true in time
```

File: integration-cli/docker_cli_run_test.go [hunk 13]
```diff
@@ -2268,11 +2473,14 @@ func (s *DockerSuite) TestRunSetDefaultRestartPolicy(c *check.C) {
 }
 
 func (s *DockerSuite) TestRunRestartMaxRetries(c *check.C) {
-	testRequires(c, DaemonIsLinux)
 	out, _ := dockerCmd(c, "run", "-d", "--restart=on-failure:3", "busybox", "false")
+	timeout := 10 * time.Second
```
Comment (calavera) [inline integration-cli/docker_cli_run_test.go]: this value is translated to `time.Duration` inside the `waitInspect` function. I think the problem with the flaky test is that the double translation causes that value to not be 10/45 seconds, I've seen similar issues in the past

Comment (jhowardmsft) [inline integration-cli/docker_cli_run_test.go]: Updating waitInspect function to have after := time.After(timeout) as per IRC.

**Comment by jhowardmsft** [comment]
Good call @calavera  on the tags until Jenkins is right :)

**Comment by cpuguy83** [comment]
LGTM
